### PR TITLE
feat(parser): Initial support for parsing flags in JS

### PR DESCRIPTION
**Pull request by nicojs**
+ Initial support for parsing flags in JS
+ Add and update JS parsing rules that are affected by the Unicode flag `u`. Specifically, with the `u` flag:
  * `\u` marks the start of a Unicode escape sequence, otherwise it matches the character `u`
  * `\p{UnicodeProperty}` matches a Unicode property escape, otherwise it matches the character `p` followed by the string `{UnicodeProperty}`
  * a character identity escape `\x` is only valid if `x` is a syntax character `^ $ \ . * + ? ( ) [ ] { } |` or `/`. Without the `u` flag, a character identity escape accepts other characters
+ Add and update tests for parsing rules that are affected by the Unicode flag `u`

### CODE REVIEW

File: core/src/main/scalajs/weaponregex/WeaponRegeXJS.scala [hunk 1]
```diff
@@ -36,7 +38,7 @@ object WeaponRegeXJS {
     *   A JavaScript Array of [[weaponregex.model.mutation.Mutant]] if can be parsed, or throw an exception otherwise
     */
   @JSExportTopLevel("mutate")
-  def mutate(pattern: String, options: MutationOptions = new MutationOptions): js.Array[MutantJS] = {
+  def mutate(pattern: String, flags: String, options: MutationOptions): js.Array[MutantJS] = {
```
Comment (nicojs) [inline core/src/main/scalajs/weaponregex/WeaponRegeXJS.scala:41]: Overloading doesn't exist in JS. 

Generally, overloading is mimicked with something like this:

```js
function mutate(pattern, flags, options) {
  // fix overloadings. 
  // flags is either a string or options or undefined
  // options is either options or undefined
  if (!options) {
    options = {};
  }
  if (typeof flags === 'string') {
     options.flags = flags;
  } else {
     options = flags;
  }
}
```

I'm not sure if this is supported out-of-the-box by scala-js, maybe @hugo-vrijswijk knows how this works?

Since flags are optional, it might also make sense to put them in the `MutationOptions` class. Although you might not consider it to be a "mutation option".

File: core/src/test/scala/weaponregex/parser/ParserJSTest.scala [hunk 2]
```diff
@@ -74,6 +98,93 @@ class ParserJSTest extends munit.FunSuite with ParserTest {
...
+    val pattern = "\\u20"
+    val parsedTree = Parser(pattern, "", parserFlavor).get.to[Concat]
+
+    assert(clue(parsedTree.children.head) match {
+      case QuoteChar('u', _) => true
+      case _                 => false
+    })
+
+    treeBuildTest(parsedTree, pattern)
+  }
```
Comment (ericcornelissen) [inline core/src/test/scala/weaponregex/parser/ParserJSTest.scala:111]: This test appears to be a duplicate of [the test on line 65](https://github.com/stryker-mutator/weapon-regex/pull/165/files#diff-a46660c20dc5953c0f693b59b26cb3fb7ac2e903b61bca65415af263831cd029R65):

https://github.com/stryker-mutator/weapon-regex/blob/1b75fa63965a15b6a37d795c69bd8fe4fd2052b4/core/src/test/scala/weaponregex/parser/ParserJSTest.scala#L65-L75

The only difference is the "a" in front of "character" in the test name:

```diff
-  test("Parse `\\u20` as character quotation without the Unicode flag") {
+  test("Parse `\\u20` as a character quotation without the Unicode flag") {
    val pattern = "\\u20"
    val parsedTree = Parser(pattern, "", parserFlavor).get.to[Concat]

    assert(clue(parsedTree.children.head) match {
      case QuoteChar('u', _) => true
      case _                 => false
    })

    treeBuildTest(parsedTree, pattern)
  }
```

File: core/src/main/scalajs/weaponregex/WeaponRegeXJS.scala (hunk 1 above)
Comment (ericcornelissen) [reply core/src/main/scalajs/weaponregex/WeaponRegeXJS.scala:41]: > Since flags are optional, it might also make sense to put them in the `MutationOptions` class. Although you might not consider it to be a "mutation option".

I wouldn't consider it a mutation option so I think that'd be bad API design. I think the current API is the best available.

An alternative would be to make the first argument a class as well (with `pattern` and optional `flags` fields). This might be more sustainable, allowing for further extension of the class in the future. Unfortunately that'd still require method overloading (or a breaking change).

File: core/src/main/scala/weaponregex/parser/Parser.scala [hunk 3]
```diff
@@ -12,26 +12,41 @@ import scala.util.{Failure, Success, Try}
...
 /** The based abstract parser
   * @param pattern
   *   The regex pattern to be parsed
+  * @param flags
+  *   The regex flags to be used
   * @note
   *   The parsing rules methods inside this class is created based on the defined grammar
   */
-abstract class Parser(val pattern: String) {
+abstract class Parser(val pattern: String, val flags: String) {
```
Comment (hugo-vrijswijk) [inline core/src/main/scala/weaponregex/parser/Parser.scala:47]: Since flags is only something that is relevant to the ParserJS, can the parameter be moved to only be added there? ParserJVM will never have a flags as it doesn't exist

File: core/src/main/scalajs/weaponregex/WeaponRegeXJS.scala (hunk 1 above)
Comment (hugo-vrijswijk) [reply core/src/main/scalajs/weaponregex/WeaponRegeXJS.scala:41]: I would expect that two exported functions of the same name would create a linking error in Scala.js, but that doesn't seem to be the case. 
Suggestion: either create a `mutateWithFlags`, or have the flags as a _third_ optional argument after options. That wouldn't break anything (at least source compatibility, but js doesn't have binary compatibility 🤠)

File: core/src/test/scala/weaponregex/parser/ParserTest.scala [hunk 4]
```diff
@@ -22,15 +22,17 @@ trait ParserTest {
...
+  def parseErrorTest(pattern: String, flags: String): Unit = {
+    val parsedTree = Parser(pattern, flags, parserFlavor)
 
     assert(clue(parsedTree) match {
       case Failure(exception: RuntimeException) => exception.getMessage.startsWith("[Error] Parser:")
       case _                                    => false
     })
   }
 
+  def parseErrorTest(pattern: String): Unit = parseErrorTest(pattern, "")
```
Comment (hugo-vrijswijk) [inline core/src/test/scala/weaponregex/parser/ParserTest.scala:34]: This can also be a default argument for flags in the other function

File: core/src/main/scalajs/weaponregex/WeaponRegeXJS.scala (hunk 1 above)
Comment (nicojs) [reply core/src/main/scalajs/weaponregex/WeaponRegeXJS.scala:41]: I did a small test. This:

```scala
 class MutationOptions(
      val mutationLevels: js.Array[Int] = null
  ) extends js.Object
object Parser {
  @JSExportTopLevel("mutate") 
  def mutate(pattern: String, flags: String, options: MutationOptions): Unit = {
  }
  @JSExportTopLevel("mutate") 
  def mutate(pattern: String, options: MutationOptions): Unit = {
  }
}
```

Gets compiled into:

```js
mutate = (function(arg, arg$2, ...rest) {
  switch ($uI(rest.length)) {
    case 0: {
      var prep0 = $as_T(arg);
      var prep1 = arg$2;
      $m_LParser$().mutate__T__LMutationOptions__V(prep0, prep1);
      return (void 0);
      break
    }
    case 1: {
      var prep0$2 = $as_T(arg);
      var prep1$2 = $as_T(arg$2);
      var prep2 = rest[0];
      $m_LParser$().mutate__T__T__LMutationOptions__V(prep0$2, prep1$2, prep2);
      return (void 0);
      break
    }
    default: {
      throw new TypeError("No matching overload")
    }
  }
});
```

So overloads seem to be supported just fine.

I wouldn't worry too much about breaking changes, since we might want to bundle this in a major release anyhow because of https://github.com/stryker-mutator/weapon-regex/issues/166

File: core/src/test/scala/weaponregex/parser/ParserJSTest.scala (hunk 2 above)
Comment (Nhaajt) [reply core/src/test/scala/weaponregex/parser/ParserJSTest.scala:111]: My bad 😂

File: core/src/main/scala/weaponregex/parser/Parser.scala (hunk 3 above)
Comment (Nhaajt) [reply core/src/main/scala/weaponregex/parser/Parser.scala:47]: The `Pattern` in Java somehow seems to have support for “global” flags ([reference](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/regex/Pattern.html#compile(java.lang.String,int))). I also checked and many online regex testers for Java also have global flags support.

File: core/src/test/scala/weaponregex/parser/ParserTest.scala (hunk 4 above)
Comment (Nhaajt) [reply core/src/test/scala/weaponregex/parser/ParserTest.scala:34]: Of course it should be 🤦‍♂️

File: core/src/main/scalajs/weaponregex/WeaponRegeXJS.scala (hunk 1 above)
Comment (Nhaajt) [reply core/src/main/scalajs/weaponregex/WeaponRegeXJS.scala:41]: I also had the same doubts before, but Scala.js did some clever work behind the scenes so “overloads” seems to work just fine in JS. Anyway, how do you want the API, since Stryker is virtually the only user of this project 😂

Comment (nicojs) [reply core/src/main/scalajs/weaponregex/WeaponRegeXJS.scala:41]: In that case, let's do what fits best with your design. Flags as a second argument are fine by me!

Btw, this package has ~50k downloads per week, isn't that amazing!!! 💙

File: core/src/main/scala/weaponregex/parser/Parser.scala (hunk 3 above)
Comment (hugo-vrijswijk) [reply core/src/main/scala/weaponregex/parser/Parser.scala:47]: Are they relevant to mutating? Especially since they're not used and have a very different implementation compared to js, I think it's cleaner to keep it in ParserJS. If we ever need to add flags support for jvm we can add them to ParserJVM

File: core/src/main/scala/weaponregex/parser/Parser.scala [hunk 5]
```diff
@@ -187,18 +218,30 @@ abstract class Parser(val pattern: String) {
...
     *   [[weaponregex.model.regextree.MetaChar]] tree node
     * @example
-    *   `"\x{0123}"`
+    *   `"\ x{0123}"` or `"\ u{0123}"`
+    * @see
+    *   [[weaponregex.parser.Parser.codePointEscChar]]
     */
-  def charHexBrace[A: P]: P[MetaChar] = Indexed("""\x{""" ~ CharIn("0-9a-zA-Z").rep(1).! ~ "}")
-    .map { case (loc, hexDigits) => MetaChar("x{" + hexDigits + "}", loc) }
+  def charCodePoint[A: P]: P[MetaChar] = Indexed(s"\\$codePointEscChar" ~ ("{" ~ hexDigit.rep(1) ~ "}").!)
```
Comment (ericcornelissen) [inline core/src/main/scala/weaponregex/parser/Parser.scala:233]: Perhaps a bit out of scope but given that the changes of #167 got pulled in here as well:

I think there's one more small problem with this part of the parser. In particular, from my testing on [regex101.com](https://regex101.com/) with JVM (using `"\x{h..h}"`) and JS (using `/\u{h..h}/u`) I found that it only accepts up-to a 5-digit (hexadecimal) number. So, I think this should be:

```suggestion
  def charCodePoint[A: P]: P[MetaChar] = Indexed(s"\\$codePointEscChar" ~ ("{" ~ hexDigit.rep(min=1, max=5) ~ "}").!)
```

However, both JVM and JS seem to be fine with any number of leading zeros, so technically it should be something like:

```suggestion
  def charCodePoint[A: P]: P[MetaChar] = Indexed(s"\\$codePointEscChar" ~ ("{" ~ "0".rep(min=0) ~ hexDigit.rep(min=1, max=5) ~ "}").!)
```

Comment (Nhaajt) [reply core/src/main/scala/weaponregex/parser/Parser.scala]: Thanks for the bug report 👍 but technically it should accept upto 6 hex digits (without leading zeros) since a code point can be anything `0 <= codepoint <= 0x10FFFF` ([JS reference](https://tc39.es/ecma262/multipage/ecmascript-language-lexical-grammar.html#prod-CodePoint), [Java reference](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/regex/Pattern.html#characters))

File: core/src/main/scala/weaponregex/parser/Parser.scala (hunk 3 above)
Comment (Nhaajt) [reply core/src/main/scala/weaponregex/parser/Parser.scala:47]: Well, mutating is affected by the parsed result and global flags do affect parsing. For example, `abc #comment` without any flags matches the whole string `abc #comment`, but with the `COMMENT` global flag, it only matches `abc`, so their parse trees are different. But from what I can tell these flags only affect the parsing in a very small number of edge cases.

However, I agree that their implementations are different, so it make sense to move the current flags support to ParserJS

File: core/src/main/scalajs/weaponregex/WeaponRegeXJS.scala (hunk 1 above)
Comment (hugo-vrijswijk) [inline core/src/main/scalajs/weaponregex/WeaponRegeXJS.scala:41]: Is this change in API on purpose? These no longer have default arguments

Comment (Nhaajt) [reply core/src/main/scalajs/weaponregex/WeaponRegeXJS.scala:41]: git diff got confused 😵 the exact same old API method should just be right below this one.

Comment (hugo-vrijswijk) [reply core/src/main/scalajs/weaponregex/WeaponRegeXJS.scala:41]: Ah I see. Could we still have the `MutationOptions` as a default arg? And I think preferably the flags as well

Comment (Nhaajt) [reply core/src/main/scalajs/weaponregex/WeaponRegeXJS.scala:41]: Scala doesn’t like multiple overloaded alternatives of method with default arguments, so I chose to preserve the old API. Plus, I think it’s logical for the user to explicitly define the regex flavor in mutation options when they want to use String flags, since not all flavors implement flags as a String.

File: core/src/main/scala/weaponregex/parser/Parser.scala [hunk 6]
```diff
@@ -12,17 +13,30 @@ import scala.util.{Failure, Success, Try}
...
+    *   The regex pattern to be parsed
+    * @param flags
+    *   The regex flags to be used
+    * @return
+    *   A `Success` of parsed [[weaponregex.model.regextree.RegexTree]] if can be parsed, a `Failure` otherwise
+    */
+  def apply(pattern: String, flags: String, flavor: ParserFlavor): Try[RegexTree] =
+    flavor match {
+      case ParserFlavorJVM =>
+        if (flags.nonEmpty) Failure(new RuntimeException(ErrorMessage.jvmWithStringFlags))
```
Comment (hugo-vrijswijk) [inline core/src/main/scala/weaponregex/parser/Parser.scala:27]: This should be an `IllegalArgumentException` I think

File: core/src/main/scala/weaponregex/parser/Parser.scala [hunk 7]
```diff
@@ -12,17 +12,31 @@ import scala.util.{Failure, Success, Try}
...
     *   The regex pattern to be parsed
     * @return
     *   A `Success` of parsed [[weaponregex.model.regextree.RegexTree]] if can be parsed, a `Failure` otherwise
     */
-  def apply(pattern: String, flavor: ParserFlavor = ParserFlavorJVM): Try[RegexTree] = flavor match {
-    case ParserFlavorJVM => new ParserJVM(pattern).parse
-    case ParserFlavorJS  => new ParserJS(pattern).parse
-    case _               => Failure(new RuntimeException("[Error] Parser: Unsupported regex flavor"))
-  }
+  def apply(pattern: String, flavor: ParserFlavor = ParserFlavorJVM): Try[RegexTree] = apply(pattern, "", flavor)
```
[the remark this answers is not in the document]
Comment (hugo-vrijswijk) [reply core/src/main/scala/weaponregex/parser/Parser.scala:39]: It would signify that flags are optional. A required argument, but filled with an empty string is less clear and makes it seem like a pattern always has flags (but just empty ones), which isn't the case. Especially since it's not implemented for all parser flavours

File: core/src/main/scalajs/weaponregex/WeaponRegeXJS.scala [hunk 8]
```diff
@@ -38,7 +38,7 @@ object WeaponRegeXJS {
     *   A JavaScript Array of [[weaponregex.model.mutation.Mutant]] if can be parsed, or throw an exception otherwise
     */
   @JSExportTopLevel("mutate")
-  def mutate(pattern: String, flags: String, options: MutationOptions): js.Array[MutantJS] = {
+  def mutate(pattern: String, flags: js.UndefOr[String], options: MutationOptions): js.Array[MutantJS] = {
```
Comment (hugo-vrijswijk) [inline core/src/main/scalajs/weaponregex/WeaponRegeXJS.scala:41]: Should de options be required too? If not then the 'default' API might be a bit awkward:
```js
wrx.mutate('\\u{20}', undefined, {});
// VS
wrx.mutate('\\u{20}');
```

Comment (hugo-vrijswijk) [reply core/src/main/scalajs/weaponregex/WeaponRegeXJS.scala:41]: I just realized you already answered this question twice 😅. It's because of the overloads. I guess in a future version when the deprecated export is removed we can make `options` optional

**Review by hugo-vrijswijk — APPROVED** [review]
Sorry I didn't get back to this for a while. I thought I was waiting on you 🙈. Feel free to merge this!

**Comment by Nhaajt** [comment]
> Sorry I didn't get back to this for a while. I thought I was waiting on you 🙈. Feel free to merge this!

Perfect 👌
